tui/tui.go

```diff
@@ -0,0 +1,293 @@
+package tui
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"sync"
+	"syscall"
+
+	tui "github.com/marcusolsson/tui-go"
+	"github.com/samsarahq/taskrunner"
+)
+
+func Option(r *taskrunner.Runtime) {
+	ui := New()
+	r.OnStart(func(ctx context.Context, executor *taskrunner.Executor) error {
+		ui.executor = executor
+		ui.tabWidget.executor = executor
+		ui.Run(ctx, executor)
+		return nil
+	})
+	r.Subscribe(ui.Subscribe)
+}
+
+type uiTab struct {
+	task  *taskrunner.Task
+	label *Label
+	view  *logList
+}
+
+type tabWidget struct {
+	sync.RWMutex
+	executor       *taskrunner.Executor
+	tabIndexByTask map[*taskrunner.Task]int
+	*tui.Box
+	bar      *tui.Box
+	tabs     []*uiTab
+	selected int
+}
+
+func (t *tabWidget) tab(task *taskrunner.Task) *uiTab {
+	index, ok := t.tabIndexByTask[task]
+	if !ok {
+		return nil
+	}
+
+	return t.tabs[index]
+}
+
+type UI struct {
+	tui.UI
+	executor  *taskrunner.Executor
+	done      bool
+	tabWidget *tabWidget
+}
+
+func clamp(n, min, max int) int {
+	if n < min {
+		return max
+	}
+	if n > max {
+		return min
+	}
+	return n
+}
+
+type Label struct {
+	*tui.Box
+	Child *tui.Label
+	Style string
+}
+
+func (l *Label) Draw(p *tui.Painter) {
+	p.WithStyle(l.Style, l.Box.Draw)
+}
+
+func (l *Label) SetStyleName(style string) {
+	l.Style = style
+}
+
+func newLabel(name string) *Label {
+	label := tui.NewLabel(name)
+	return &Label{
+		Box: tui.NewHBox(
+			tui.NewSpacer(),
+			tui.NewPadder(1, 0, label),
+		),
+		Child: label,
+	}
+}
+
+func (t *tabWidget) OnKeyEvent(ev tui.KeyEvent) {
+	switch ev.Rune {
+	case 'j':
+		t.Next()
+	case 'k':
+		t.Previous()
+	case 'r':
+		go func(e *taskrunner.Executor) {
+			e.Invalidate(t.Current().task, taskrunner.UserRestart{})
+		}(t.executor)
+	}
+	switch ev.Key {
+	case tui.KeyTab:
+		t.Next()
+	case 278:
+		t.Previous()
+	case tui.KeyCtrlU, tui.KeyDown:
+		t.ScrollDown()
+	case tui.KeyCtrlD, tui.KeyUp:
+		t.ScrollUp()
+	}
+	t.Box.OnKeyEvent(ev)
+}
+
+func (t *tabWidget) style() {
+	for i := 0; i < len(t.tabs); i++ {
+		if i == t.selected {
+			t.tabs[i].label.SetStyleName("tab-selected")
+			continue
+		}
+		t.tabs[i].label.SetStyleName("tab")
+	}
+}
+
+func (t *tabWidget) setView(view tui.Widget) {
+	t.Box.Remove(1)
+	t.Box.Append(view)
+}
+
+func (t *tabWidget) Select(selection int) {
+	t.selected = clamp(selection, 0, len(t.tabs)-1)
+	t.style()
+	t.setView(t.Current().view)
+}
+func (t *tabWidget) Current() *uiTab { return t.tabs[t.selected] }
+func (t *tabWidget) Next()           { t.Select(t.selected + 1) }
+func (t *tabWidget) Previous()       { t.Select(t.selected - 1) }
+func (t *tabWidget) ScrollUp() {
+	t.Current().view.Scroll(0, +5)
+}
+func (t *tabWidget) ScrollDown() {
+	t.Current().view.Scroll(0, -5)
+}
+
+func New() *UI {
```

Comment:
The ordering of functions in this file make it a bit difficult to follow what's going on.  I don't think there's a silver bullet when it comes to ordering functions/definitions in a file, but, maybe we can try the newspaper code structure ([best link I could find by quickly](https://blog.kentcdodds.com/newspaper-code-structure-8b4616af9539)).
Essentially we order the structs/functions in the order that they are used.  When it comes to this file, the order of functions would be something like:

Option
UI struct def
UI New func
UI Run func
UI Quit func
UI Subscribe Func
TabWidget Struct Def
...

Reply:
👍 Do you think it'd help if I separate things into files at all? This code is definitely on the messier end

Reply:
LMK if it's easier to read with the split